substrate/frame/election-provider-multi-block/src/signed/mod.rs

```diff
@@ -0,0 +1,856 @@
+				let _ = <T::Verifier as AsyncVerifier>::start().defensive();
+			};
+
+			if crate::Pallet::<T>::current_phase().is_unsigned_open_at(now) {
+				sublog!(debug, "signed", "signed validation phase ended, signaling the verifier to stop async verifiacton.");
+				<T::Verifier as AsyncVerifier>::stop();
+			}
+
+			Weight::default()
+		}
+	}
+}
+
+impl<T: Config> SolutionDataProvider for Pallet<T> {
+	type Solution = SolutionOf<T::MinerConfig>;
+
+	/// Returns a paged solution of the *best* solution in the queue.
+	fn get_paged_solution(page: PageIndex) -> Option<Self::Solution> {
+		let round = crate::Pallet::<T>::current_round();
+
+		Submissions::<T>::leader(round).map(|(who, _score)| {
+			sublog!(info, "signed", "returning page {} of leader's {:?} solution", page, who);
+			Submissions::<T>::page_submission_for(round, who, page).unwrap_or_default()
+		})
+	}
+
+	/// Returns the score of the *best* solution in the queueu.
+	fn get_score() -> Option<ElectionScore> {
+		let round = crate::Pallet::<T>::current_round();
+		Submissions::<T>::leader(round).map(|(_who, score)| score)
+	}
+
+	/// Called by an external entity to report a verification result of the current *best*
+	/// solution.
+	///
+	/// If the verification is rejected, update the leader's metadata to be slashed (i.e. set
+	/// release strategy to [`ReleaseStrategy::BurnAll`] in the leader's metadata). If successful
+	/// (represented by the variant [``VerificationResult::Queued]), reward the submitter and
+	/// signal the verifier to stop the async election verification.
+	fn report_result(result: VerificationResult) {
+		let round = crate::Pallet::<T>::current_round();
+
+		let (leader, mut metadata) =
+			if let Some((leader, maybe_metadata)) = Submissions::<T>::take_leader_score(round) {
+				let metadata = match maybe_metadata {
+					Some(m) => m,
+					None => {
+						defensive!("unexpected: leader with inconsistent data (no metadata).");
+						return;
+					},
+				};
+				(leader, metadata)
+			} else {
+				// TODO(gpestana): turn into defensive.
+				sublog!(error, "signed", "unexpected: leader called without active submissions.");
+				return
+			};
+
+		match result {
+			VerificationResult::Queued => {
+				// solution was accepted by the verifier, reward leader and stop async
+				// verification.
+				// TODO(gpestana): think better about the reward minting process -- should we keep
+				// a pot for rewards instead of minting it in staking?
+				let _ = T::Currency::mint_into(&leader, T::Reward::get()).defensive();
+				let _ = <T::Verifier as AsyncVerifier>::stop();
+			},
+			VerificationResult::Rejected | VerificationResult::DataUnavailable => {
+				// updates metadata release strategy so that all the deposit is burned when the
+				// leader's data is cleared.
+				metadata.release_strategy = ReleaseStrategy::BurnAll;
+				Submissions::<T>::set_metadata(round, &leader, metadata);
```

Comment:
This will fail because set_metadata checks for the leader to be present in SortedScores but he (that entry) was already removed in line 828 by Submissions::<T>::take_leader_score(round).